src/proto/factories/optimizer_factory.cpp

```diff
@@ -26,69 +26,66 @@
 
 #include "lbann/proto/factories.hpp"
 
+#include "lbann/optimizers/optimizer.hpp"
+
 #include "lbann/optimizers/adagrad.hpp"
 #include "lbann/optimizers/adam.hpp"
 #include "lbann/optimizers/hypergradient_adam.hpp"
-#include "lbann/optimizers/optimizer.hpp"
 #include "lbann/optimizers/rmsprop.hpp"
 #include "lbann/optimizers/sgd.hpp"
 
+#include "lbann/proto/helpers.hpp"
+#include "lbann/utils/factory.hpp"
+
 #include <optimizers.pb.h>
 
 namespace lbann {
 namespace proto {
+namespace {
 
-optimizer* construct_optimizer(lbann_comm* comm,
-                               const lbann_data::Optimizer& proto_opt) {
-
-  // Stochastic gradient descent
-  if (proto_opt.has_sgd()) {
-    const auto& params = proto_opt.sgd();
-    return new sgd(comm,
-                   params.learn_rate(),
-                   params.momentum(),
-                   params.nesterov());
-  }
+using factory_type = lbann::generic_factory<
+  lbann::optimizer,
+  std::string,
+  generate_builder_type<lbann::optimizer,
+                        google::protobuf::Message const&,
+                        lbann_comm*>,
+  default_key_error_policy>;
 
-  // AdaGrad
-  if (proto_opt.has_adagrad()) {
-    const auto& params = proto_opt.adagrad();
-    return new adagrad(comm, params.learn_rate(), params.eps());
-  }
+void register_default_builders(factory_type& factory)
+{
```

Comment:
LBANN style is to put the brace on the same line.

Reply:
All fixed.